src/clj/puppetlabs/puppetserver/certificate_authority.clj

```diff
@@ -954,6 +954,23 @@
       {:kind :invalid-signature
        :msg (tru "CSR contains a public key that does not correspond to the signing key")})))
 
+(schema/defn ensure-no-authorization-extensions!
+  "Throws an exception if the CSR contains authorization exceptions. These
+  extensions need to be signed using the puppet cert tool. This may change in
+  the future, but for now, it's too risky that certificates with authentication
+  extensions could be signed unintentionally."
+  [csr :- CertificateRequest]
+  (let [extensions (utils/get-extensions csr)]
+    (doseq [extension extensions]
+      (when (utils/subtree-of? ppAuthCertExt (:oid extension))
+        (sling/throw+
+         {:kind :disallowed-extension
+          :msg (format "%s %s"
+                (trs "CSR ''{0}'' contains an authorization extension, which is disallowed." csr)
+                (trs
+                 "Use `puppet cert --allow-authorization-extensions sign {0}` to sign this request."
+                 (get-csr-subject csr)))})))))
```

Comment:
There's a tension between providing context with long strings vs the readability hit of not being able to break translated strings into multiple lines and saving costs by breaking out repeated strings. Breaking on sentence boundary is the recommended way to solve that tension.

We could do something like: `(trs "CSR ''{0}'' contains an authorization extensions, which is disallowed. Use 'puppet cert --allow-authorization-extensions sign {0} to sign this request." csr)`

But in the code it would all have to be one line. I think Nathaniel's code is more readable, and it also allows us to re-use the `"Use 'puppet cert --allow-authorization-extensions sign {0}' to sign this request."` elsewhere.

IIRC, we want to *not* translate strings that are either beyond our control (eg. certnames) or won't be translated for the user (eg. cli commands). I don't know if breaking out the cli invocation would also make the `Use...` string re-usable between authorization extensions and alt names? If we're not using similar word choice between the two responses,  since they have such similar workflows, should we?